### PR TITLE
Prefetch buckets when scanning over flow table

### PR DESCRIPTION
**Pull request by mengxiang0811**

### CODE REVIEW

File: gk/main.c [hunk 1]
```diff
@@ -2271,6 +2267,24 @@ gk_proc(void *arg)
...
+			instance, gk_conf);
+
+		process_pkts_back(port_back, port_front,
+			rx_queue_back, tx_queue_front, lcore,
+			&back_num_pkts, back_icmp_bufs, instance, gk_conf);
+
+		if (fe != NULL) {
+			flow_hash_val = rss_ip_flow_hf(&fe->flow, 0, 0);
+			rte_hash_prefetch_buckets(instance->ip_flow_hash_table,
+				flow_hash_val);
```
Comment (AltraMayor) [inline gk/main.c]: We should only calculate the hash and issue these prefetches if `fe` is in use and expired. In other words, we no longer need `gk_flow_tbl_entry_scan()`; bring the code of `gk_flow_tbl_entry_scan()` into `gk_proc()`.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: gk/main.c [hunk 2]
```diff
@@ -2271,6 +2254,24 @@ gk_proc(void *arg)
...
+
+		process_pkts_back(port_back, port_front,
+			rx_queue_back, tx_queue_front, lcore,
+			&back_num_pkts, back_icmp_bufs, instance, gk_conf);
+
+		if (fe != NULL && fe->in_use && is_flow_expired(fe, rte_rdtsc())) {
+			flow_hash_val = rss_ip_flow_hf(&fe->flow, 0, 0);
+			rte_hash_prefetch_buckets(instance->ip_flow_hash_table,
+				flow_hash_val);
 		}
```
Comment (AltraMayor) [inline gk/main.c]: 1. Any hash value is valid, including `-1`, just add an `else` clause and `fe = NULL;` there.
2. Don't initialize `flow_hash_val`.

File: gk/main.c [hunk 3]
```diff
@@ -2281,14 +2282,14 @@ gk_proc(void *arg)
 
 		process_cmds_from_mailbox(instance, gk_conf);
 
-		if (iter_count >= scan_iter) {
-			if (gk_flow_tbl_entry_scan(fe, instance) &&
-					instance->num_scan_del > 0)
-				instance->num_scan_del--;
+		if (flow_hash_val != (uint32_t)-1) {
```
Comment (AltraMayor) [inline gk/main.c]: Return test `fe != NULL` here.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

**Comment by AltraMayor** [comment]
This pull request is ready for an evaluation.

File: gk/main.c [hunk 4]
```diff
@@ -2271,7 +2254,26 @@ gk_proc(void *arg)
...
+			lcore, &front_num_pkts, front_icmp_bufs,
+			instance, gk_conf);
+
+		process_pkts_back(port_back, port_front,
+			rx_queue_back, tx_queue_front, lcore,
+			&back_num_pkts, back_icmp_bufs, instance, gk_conf);
+
+		if (fe != NULL && fe->in_use && is_flow_expired(fe, rte_rdtsc())) {
+			flow_hash_val = rss_ip_flow_hf(&fe->flow, 0, 0);
+			rte_hash_prefetch_buckets(instance->ip_flow_hash_table,
```
Comment (AltraMayor) [inline gk/main.c]: Use the non-temporal version of `rte_hash_prefetch_buckets()`.

Comment (mengxiang0811) [reply gk/main.c]: Done.

**Comment by mengxiang0811** [comment]
Below is the performance of the master branch:

| lcores | table size | GK Mpps rcvd (0%) | GK Mpps rcvd (50%) | GK Mpps rcvd (99%) | GK Mpps rcvd (mean) | cli Mpps sent (0%) | cli Mpps sent (50%) | cli Mpps sent (99%) | cli Mpps sent (mean) |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  1  |  2^15  |       5.74        |        5.94        |        6.56        |        6.08         |       8.91        |        9.97        |        11.07        |         9.95         | 
|  1  |  2^20  |       4.68        |        4.77        |        5.16        |        4.87         |       9.26        |        10.38        |        11.56        |         10.49         | 
|  1  |  2^22  |       3.61        |        3.98        |        4.03        |        3.87         |       9.04        |        10.34        |        11.38        |         10.26         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  2  |  2^15  |       7.11        |        8.07        |        8.7        |        7.99         |       9.04        |        10.38        |        10.88        |         10.27         | 
|  2  |  2^20  |       6.25        |        6.81        |        7.76        |        6.91         |       9.11        |        10.06        |        11.25        |         10.13         | 
|  2  |  2^22  |       2.31        |        3.12        |        5.1        |        3.43         |       9.13        |        10.46        |        10.8        |         10.34         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  3  |  2^15  |       8.37        |        9.2        |        9.55        |        9.09         |       8.95        |        10.02        |        10.77        |         9.95         | 
|  3  |  2^20  |       0.85        |        0.91        |        1.19        |        0.95         |       8.7        |        9.41        |        9.97        |         9.47         | 
|  3  |  2^22  |       1.14        |        1.36        |        1.5        |        1.32         |       8.65        |        9.69        |        10.35        |         9.72         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|

Here is the performance after applying this patch:

| lcores | table size | GK Mpps rcvd (0%) | GK Mpps rcvd (50%) | GK Mpps rcvd (99%) | GK Mpps rcvd (mean) | cli Mpps sent (0%) | cli Mpps sent (50%) | cli Mpps sent (99%) | cli Mpps sent (mean) |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  1  |  2^15  |       4.72        |        5.8        |        6.58        |        5.71         |       10.48        |        11.41        |        11.87        |         11.39         | 
|  1  |  2^20  |       3.76        |        3.76        |        5.52        |        4.36         |       10.7        |        12.03        |        12.52        |         12.03         | 
|  1  |  2^22  |       2.97        |        3.84        |        3.95        |        3.58         |       10.91        |        11.98        |        12.31        |         11.9         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  2  |  2^15  |       8.54        |        9.42        |        9.7        |        9.27         |       10.79        |        11.48        |        12.31        |         11.51         | 
|  2  |  2^20  |       4.32        |        8.06        |        8.51        |        7.28         |       10.53        |        11.16        |        11.98        |         11.2         | 
|  2  |  2^22  |       5.67        |        6.09        |        7.58        |        6.37         |       10.45        |        11.42        |        11.92        |         11.43         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  3  |  2^15  |       7.29        |        10.69        |        11.2        |        10.16         |       10.56        |        11.11        |        11.87        |         11.11         | 
|  3  |  2^20  |       6.48        |        8.78        |        9.58        |        8.26         |       10.41        |        11.34        |        11.79        |         11.31         | 
|  3  |  2^22  |       1.56        |        1.83        |        2.0        |        1.77         |       11.08        |        11.67        |        12.07        |         11.67         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|

**Pull request merged**